betterproto/plugin.py

```diff
@@ -174,266 +212,253 @@ def generate_code(request, response):
+                    "repeated": repeated,
+                    "packed": packed,
+                    "one_of": one_of,
+                }
+            )
+
+        template_data["messages"].append(item_data)
+        global_messages.append(item_data)
+    elif isinstance(item, EnumDescriptorProto):
+        item_data.update(
+            {
+                "type": "Enum",
+                "comment": get_comment(proto_file, path),
+                "entries": [
+                    {
+                        "name": v.name,
+                        "value": v.number,
+                        "comment": get_comment(proto_file, path + [2, i]),
+                    }
+                    for i, v in enumerate(item.value)
+                ],
+            }
+        )
+        template_data["enums"].append(item_data)
+
+
+def read_service(i, package, proto_file, service, template_data, global_messages):
+    data = {
+        "name": service.name,
+        "py_name": stringcase.pascalcase(service.name),
+        "comment": get_comment(proto_file, [6, i]),
+        "methods": [],
+    }
+    for j, method in enumerate(service.method):
+        if method.client_streaming:
+            raise NotImplementedError("Client streaming not yet supported")
+        sys.stderr.write(f'  add {method.name}\n')
+
+        input_message = None
+        input_type = get_ref_type(
+            package, template_data["imports"], method.input_type
+        ).strip('"')
+        for msg in global_messages:
+            if msg["full_name"] == method.input_type.lstrip('.'):
+                input_message = msg
+                for field in msg["properties"]:
+                    if field["zero"] == "None":
+                        template_data["typing_imports"].add("Optional")
+                break
+
+        if not input_message:
+            sys.stderr.write(f"no input message found for {input_type} / {method}\n")
+            sys.stderr.write(f" {global_messages}\n")
+            sys.stderr.write(f"\n")
+
+        method_data = {
+            "name": method.name,
+            "py_name": stringcase.snakecase(method.name),
+            "comment": get_comment(proto_file, [6, i, 2, j], indent=8),
+            "route": f"/{package}.{service.name}/{method.name}",
+            "input": get_ref_type(
+                package, template_data["imports"], method.input_type
+            ).strip('"'),
+            "input_message": input_message,
+            "output": get_ref_type(
+                package, template_data["imports"], method.output_type, unwrap=False
+            ).strip('"'),
+            "client_streaming": method.client_streaming,
+            "server_streaming": method.server_streaming,
+        }
+        # sys.stderr.write(f'METHOD: {method_data}\n')
+        data["methods"].append(method_data)
+
+        if method.server_streaming:
+            template_data["typing_imports"].add("AsyncGenerator")
+    template_data["services"].append(data)
+
+
```

Comment:
The diff makes it look like this is all new. Is it fair to assume none of it really is?

Reply:
Yes. Its because I've moved some code into functions, and the changed indentation has made it look new.

However, I did make an important change on line 427

     if msg["full_name"] == method.input_type.lstrip('.'):

to compare with the full name, rather than part of the name, so that all messages can be matched consistently.